keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,964 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposed enhancement adds a new `VolumeSource` to Kubernetes that supports OCI images and/or OCI artifacts.
+This allows users to package files and share them among containers in a pod without including them in the main image,
+thereby reducing vulnerabilities and simplifying image creation.
+
+While OCI images are well-supported by Kubernetes and CRI,
+extending support to OCI artifacts involves recognizing additional media types, implementing custom lifecycle management,
+and ensuring appropriate validation and security measures.
+
+## Motivation
+
+Supporting OCI images and artifacts directly as a `VolumeSource` allows
+Kubernetes to focus on OCI standards as well as allows to store and distribute
+any content using OCI registries. This allows the project to grow into use cases
+which go beyond running particular images.
+
+### Goals
+
+- Introduce a new `VolumeSource` type that allows mounting OCI images and/or artifacts.
+- Simplify the process of sharing files among containers in a pod.
+
+### Non-Goals
+
+- This proposal does not aim to replace existing `VolumeSource` types.
+- This proposal does not address other use cases for OCI objects beyond file sharing among containers in a pod.
+
+## Proposal
+
+We propose to add a new `VolumeSource` that supports OCI images and/or artifacts. This `VolumeSource` will allow users to mount an OCI object
+directly into a pod, making the files within the image accessible to the containers without the need to include them in the main image and to be able to host them in OCI compatible registries.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Kubernetes user, I want to share a configuration file among multiple containers in a pod without including the file in my main image, so that I can
+minimize security risks and image size. 
+
+Beside that, I want:
+- to package this file in an OCI object to take advantage of OCI distribution.
+- the image to be downloaded with the same credentials that kubelet using for other images.
+- to be able to use image pull secrets when downloading the image if an image is from the registry that requires image pull secrets.
+
+#### Story 2
+
+As a DevOps engineer, I want to package and distribute binary artifacts using OCI images and mount them directly into my Kubernetes pods,
```

Comment:
I worry this would get overused instead of creating an image with a single added layer.

Reply:
nod.. some of the language has been artifacts/images.. 

```suggestion
As a DevOps engineer, I want to package and distribute binary artifacts using OCI image and distribution specification standards and mount them directly into my Kubernetes pods,
```

Though it's not clear if that means we need another story for docker manifest types.. maybe best to just say Docker/OCI here.

Reply:
> I worry this would get overused instead of creating an image with a single added layer.

I think that is the advantage here: Sharing single files/layers across different images without having to sync the base layers. it's a workaround on oci image spec not being smarter about shared files

Reply:
> Sharing single files/layers across different images without having to sync the base layers.

Is "sync the base layers" a reference to coping layers between registries and repositories? If the volume source and container image are in separate repositories, that could have authentication impacts (credentials able to pull an image but not the volume source).

Reply:
not quite: two different containers sharing the same file but otherwise don't share anything else have two options, share a same base layer with that file (not super idiomatic, as I think lots of non-multi-stage-built images are built starting with package installation, which in this case would be separate), or sharing them with a volume. but a volume doesn't give the utility and tooling of oci distribution spec. I find this to be the best of both worlds

Reply:
When in the same repository, a common layer can be easily added to multiple images. Much easier, and more portable, than adding a k8s specific volume.

Reply:
I may be speaking out of my depth here, but if they're not explicitly sharing a base layer, will the storage for that layer be actually shared in the containerd/cri-o level? will the sha for that layer be the same so it's not pulled twice?

Reply:
The sha for a layer can be the same across multiple images, and when that happens, most runtimes will dedup the layers, skipping pulls for content it already has.

Reply:
> I think that is the advantage here: Sharing single files/layers across different images without having to sync the base layers. it's a workaround on oci image spec not being smarter about shared files

To me this is about separating code from data.

You might also want to use a public image but with your own data. For example, you use a public image for a malware scanner and mount in a volume of private (commercial) malware signatures. This KEP gives you a way to load those signatures without baking your own combined image (which might not be allowed by the copyright on the public image), and they'll work regardless of the OS or version of the scanning software.

It'll work with other kinds of read-only data; for example, you could have some ML retraining code (eg Python) and a base model (half a terabyte of sparse matrices). Iterating on the Python shouldn't mean re-pulling the base model.

Reply:
If we're talking about combining _code_ with code, is that something we'd like to explicitly encourage?

Reply:
> If we're talking about combining code with code, is that something we'd like to explicitly encourage?

There's multiple levels to this; when talking about "code" there's actually a somewhat fuzzy line in some cases between "configuration" and "code".  But an example *strong* dividing line is as soon as you have architecture (aarch64/x86_64/s390x/...) dependent stuff it's *clearly* code (and gets close to OCI containers obviously).

Whereas I think the primary use cases this starts from are much closer to architecture-independent data.

Reply:
The story seems to have multiple shapes, but generally I do not see any blockers in the implementation proposal about them. I rephrase the story as @mikebrow suggested. 

Is there anything left open as part of this discussion?

Reply:
Hi everyone, I'm proposing to add a new MediaType to the OCI image spec that will only hold raw data files, for optimization of large files in the image, please take a look.
https://github.com/opencontainers/image-spec/pull/1197